### ISSUE TITLE
Questions about train_sc and repetition penalty

### ISSUE THREAD

**Issue by LongForCMU** [lukemelas/image-paragraph-captioning#1]
Thank you for your excellent work! It helps me a lot! 
But I am confused about the following questions:
1. The last step in README.md named "And train with self-critical: mkdir log_sc ./train_**xe**.sh" may be "mkdir log_sc ./train_**sc**.sh" I guess? And if so, the second line in "train_sc.sh" cause an error: "unrecognized arguments: --caption_model topdown". I wonder how to fix this?
2. "train_sc.sh" has set the option "--block_trigrams" to "1", but in "train.py" line 130, you use the new defined "opt={'sample_max':0}" but not the global "opt". My understanding is that the "block_trigrams" will still be 0 and you don't use Repetition Penalty during sampling. Is that right?

**Comment by arjung128** [comment]
Yeah I believe that the 2nd instance should be "./train_sc.sh" and not "./train_xe.sh."

Were you able to create a "infos_xe-best.pkl" file? When I trained the network, it didn't create a "infos_xe-best.pkl" file.

**Comment by lukemelas** [comment]
Hello! I replied to your email, but I'm also posting this here for everyone else.

Thank you for your email. I am happy to hear you are interested in the paper! 

1. Yes, it should be _train_sc_. 

2. You are correct that the repetition penalty is not used on line 130. However, it is used on the next line, 131, when calculating the CIDEr score of the greedy search output. When calculating the reward, this quantity is subtracted from the CIDEr score of the output from sampling (and then multiplied by the gradient of the log probs).  For self-critical training, the training is paragraph-level ("sequence-level" where the entire paragraph is the "sequence"). For non-self-critical training (cross-entropy), the training is token-level. 

I hope this helps!

**Comment by lukemelas** [comment]
Feel free to reopen this is you have any additional questions.

**Comment by ruotianluo** [comment]
https://github.com/lukemelas/image-paragraph-captioning/blob/master/misc/rewards.py#L41

Greedy decoding when calculating cider is not using opt as well?

**Comment by lukemelas** [comment]
Yes, this is correct, right? It corresponds to here in your current codebase: https://github.com/ruotianluo/self-critical.pytorch/blob/master/misc/loss_wrapper.py#L24

**Comment by ruotianluo** [comment]
But that results in block_trigrams to be 0 because that's the default?

**Comment by lukemelas** [comment]
Ah yes absolutely, I am so sorry! On my local copy I have `block_trigrams = opt.get('block_trigrams', 1)` in the model definition [1], as that was easier for development. When I pushed to master, I switched the default, but I should have added `opt={"block_trigrams": opt.block_trigrams }` to the line you specified! I so apologize for that, I'll fix it in master now. 

[1] https://github.com/lukemelas/image-paragraph-captioning/blob/master/models/AttModel.py#L194

**Comment by ruotianluo** [comment]
@lukemelas if you have `block_trigrams = opt.get('block_trigrams', 1)`, doesn't that mean when you sample you also use blcok trigram?

**Comment by sushant52** [comment]
"And if so, the second line in "train_sc.sh" cause an error: "unrecognized arguments: --caption_model topdown"
How do I resolve this?

**Comment by RunCode90** [comment]
@LongForCMU 
Hi! Have you got these features: parabu_fc and parabu_att? Can you share a link? Best wishes.
